### PR TITLE
Publish retained messages only after connack

### PR DESCRIPTION
**Pull request by mochi-co**
Alternative to #143 / #144

- Ensure retained messages on inherited subscriptions are only published following the Connack packet.

### CODE REVIEW

File: server.go [hunk 1]
```diff
@@ -351,6 +351,13 @@ func (s *Server) attachClient(cl *Client, listener string) error {
 		return fmt.Errorf("ack connection packet: %w", err)
 	}
 
+	// Publish any retained messages for subscriptions which already existed on client takeover.
+	if sessionPresent {
+		for _, sub := range cl.State.Subscriptions.GetAll() {
```
Comment (mochi-co) [inline server.go:356]: As the new client cannot subscribe until the connack has been issued, GetAll can only contain inherited subscriptions.

File: server_test.go [hunk 2]
```diff
@@ -445,7 +448,11 @@ func TestEstablishConnectionInheritExisting(t *testing.T) {
 		require.ErrorIs(t, v.StopCause(), packets.CodeDisconnect) // true error is disconnect
 	}
 
-	require.Equal(t, packets.TPacketData[packets.Connack].Get(packets.TConnackAcceptedSessionExists).RawBytes, <-recv)
+	connackPlusPacket := append(
+		packets.TPacketData[packets.Connack].Get(packets.TConnackAcceptedSessionExists).RawBytes,
+		packets.TPacketData[packets.Publish].Get(packets.TPublishRetain).RawBytes...,
+	)
+	require.Equal(t, connackPlusPacket, <-recv)
```
Comment (mochi-co) [inline server_test.go:455]: Ensure connack is always before retained message

**Comment by mochi-co** [comment]
@thedevop

**Comment by thedevop** [comment]
Looks good!

**Comment by mochi-co** [comment]
@thedevop - I have a concerns about the use of `true` for the existed parameter. Please check my reasoning:

- A client subscription can be inherited from a new client from an existing client.
- A new client cannot subscribe before inheriting a subscription.
- Therefore, a new client cannot have any existing subscriptions.
- 1) Therefore, the value for existed in inheritClientSession must always be true, as the subscription already existed on the client being inherited from. 
- 2)  If so, then the logic `if !existed { atomic.AddInt64(&s.Info.Subscriptions, 1) }` in inheritClientSession must also be impossible to reach, and therefore serves no purpose.

Does this sound correct?

**Comment by thedevop** [comment]
I didn't look into the Topics.Subscribe before, but you're correct, since these are inherited subscription from the same clientID, therefore it should always be existed.

**Pull request merged**